# [응용1] 다중 선형회귀 연습 - load_diabetes

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
from sklearn.datasets import load_diabetes
diab = load_diabetes()
diab

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990749, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06833155, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286131, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04688253,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452873, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00422151,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

### 데이터 수집

In [3]:
diab.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])

## 데이터 수집

In [4]:
diab.data

array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
         0.01990749, -0.01764613],
       [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
        -0.06833155, -0.09220405],
       [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
         0.00286131, -0.02593034],
       ...,
       [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
        -0.04688253,  0.01549073],
       [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
         0.04452873, -0.02593034],
       [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
        -0.00422151,  0.00306441]])

In [6]:
pd.DataFrame(diab.data)

0         1         2         3         4         5         6  \
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
..        ...       ...       ...       ...       ...       ...       ...   
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

            7         8         9  
0   -0.002592  0.019907 -0.017646  
1   -0.039493 -0.068332 -0.092204  
2   -0.002592  0.002861 -0.025930  
3    0.034309  0.022688 -0.009362  
4   -0.002592 -0.031988 -0.046641  
..        ...       ...       ...  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018114  0.044485  
439 -0.011080 -0.046883  0.015491  
440  0.026560  0.044529 -0.025930  
441 -0.039493 -0.004222  0.003064  

[442 rows x 10 columns]

In [ ]:
diab.target

array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
        69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
        68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
        87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
       259.,  53., 190., 142.,  75., 142., 155., 225.,  59., 104., 182.,
       128.,  52.,  37., 170., 170.,  61., 144.,  52., 128.,  71., 163.,
       150.,  97., 160., 178.,  48., 270., 202., 111.,  85.,  42., 170.,
       200., 252., 113., 143.,  51.,  52., 210.,  65., 141.,  55., 134.,
        42., 111.,  98., 164.,  48.,  96.,  90., 162., 150., 279.,  92.,
        83., 128., 102., 302., 198.,  95.,  53., 134., 144., 232.,  81.,
       104.,  59., 246., 297., 258., 229., 275., 281., 179., 200., 200.,
       173., 180.,  84., 121., 161.,  99., 109., 115., 268., 274., 158.,
       107.,  83., 103., 272.,  85., 280., 336., 281., 118., 317., 235.,
        60., 174., 259., 178., 128.,  96., 126., 28

### 데이터 이해

In [ ]:
diab.keys()
print(diab.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

10개의 baselines variable
6개의 혈액 검사결과
442명의 당뇨병 환장
1년후의 당뇨병 진행도를 증감을 양으로 

10개의 특성을 표준편차를 가지고 scaled 되어짐.

In [7]:
pd.options.display.float_format = '{:.4f}'.format

In [8]:
# 기초통계량
pd.DataFrame(diab.data).describe()

0        1        2        3        4        5        6        7  \
count 442.0000 442.0000 442.0000 442.0000 442.0000 442.0000 442.0000 442.0000   
mean   -0.0000   0.0000  -0.0000  -0.0000  -0.0000   0.0000  -0.0000  -0.0000   
std     0.0476   0.0476   0.0476   0.0476   0.0476   0.0476   0.0476   0.0476   
min    -0.1072  -0.0446  -0.0903  -0.1124  -0.1268  -0.1156  -0.1023  -0.0764   
25%    -0.0373  -0.0446  -0.0342  -0.0367  -0.0342  -0.0304  -0.0351  -0.0395   
50%     0.0054  -0.0446  -0.0073  -0.0057  -0.0043  -0.0038  -0.0066  -0.0026   
75%     0.0381   0.0507   0.0312   0.0356   0.0284   0.0298   0.0293   0.0343   
max     0.1107   0.0507   0.1706   0.1320   0.1539   0.1988   0.1812   0.1852   

             8        9  
count 442.0000 442.0000  
mean    0.0000   0.0000  
std     0.0476   0.0476  
min    -0.1261  -0.1378  
25%    -0.0332  -0.0332  
50%    -0.0019  -0.0011  
75%     0.0324   0.0279  
max     0.1336   0.1356

In [12]:
type(diab.data)

numpy.ndarray

In [14]:
diab_df = pd.DataFrame(data=diab.data, columns=diab.feature_names)
diab_df.head()

age     sex     bmi      bp      s1      s2      s3      s4      s5  \
0  0.0381  0.0507  0.0617  0.0219 -0.0442 -0.0348 -0.0434 -0.0026  0.0199   
1 -0.0019 -0.0446 -0.0515 -0.0263 -0.0084 -0.0192  0.0744 -0.0395 -0.0683   
2  0.0853  0.0507  0.0445 -0.0057 -0.0456 -0.0342 -0.0324 -0.0026  0.0029   
3 -0.0891 -0.0446 -0.0116 -0.0367  0.0122  0.0250 -0.0360  0.0343  0.0227   
4  0.0054 -0.0446 -0.0364  0.0219  0.0039  0.0156  0.0081 -0.0026 -0.0320   

       s6  
0 -0.0176  
1 -0.0922  
2 -0.0259  
3 -0.0094  
4 -0.0466

### 데이터 준비

In [13]:
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

def get_score(y_test, y_pred):
    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_test, y_pred)


    print(f'mse - {mse:.4f}')
    print(f'rmse - {rmse: .4f}')
    print(f'r2 - {r2: .4f}')
    return

In [16]:
x = diab.data
y = diab.target

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=27)

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  
0 -0.002592  0.019907 -0.017646  
1 -0.039493 -0.068332 -0.092204  
2 -0.002592  0.002861 -0.025930  
3  0.034309  0.022688 -0.009362  
4 -0.002592 -0.031988 -0.046641

### 모델 구축

In [ ]:
diab_df['y'] = diab.target
diab_df.head()

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6      y  
0 -0.002592  0.019907 -0.017646  151.0  
1 -0.039493 -0.068332 -0.092204   75.0  
2 -0.002592  0.002861 -0.025930  141.0  
3  0.034309  0.022688 -0.009362  206.0  
4 -0.002592 -0.031988 -0.046641  135.0

In [ ]:
diab_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     442 non-null    float64
 1   sex     442 non-null    float64
 2   bmi     442 non-null    float64
 3   bp      442 non-null    float64
 4   s1      442 non-null    float64
 5   s2      442 non-null    float64
 6   s3      442 non-null    float64
 7   s4      442 non-null    float64
 8   s5      442 non-null    float64
 9   s6      442 non-null    float64
 10  y       442 non-null    float64
dtypes: float64(11)
memory usage: 38.1 KB


In [ ]:
from sklearn.model_selection import train_test_split

x = diab_df.drop(['y'], axis=1)
y = diab_df['y']

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=27)

In [ ]:
x_train.head()

age       sex       bmi        bp        s1        s2        s3  \
169 -0.001882 -0.044642 -0.026684  0.049415  0.058973 -0.016032 -0.047082   
256 -0.049105 -0.044642  0.160855 -0.046985 -0.029088 -0.019790 -0.047082   
195  0.027178  0.050680  0.025051  0.014987  0.025950  0.048477 -0.039719   
63  -0.034575 -0.044642 -0.037463 -0.060756  0.020446  0.043466 -0.013948   
12   0.016281 -0.044642 -0.028840 -0.009113 -0.004321 -0.009769  0.044958   

           s4        s5        s6  
169  0.071210  0.133597  0.019633  
256  0.034309  0.028020  0.011349  
195  0.034309  0.007838  0.023775  
63  -0.002592 -0.030748 -0.071494  
12  -0.039493 -0.030748 -0.042499

In [ ]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

In [ ]:
reg = lr.fit(x_train, y_train)

In [ ]:
# 회귀계수가 10개
reg.coef_

array([  42.13333043, -256.72425295,  585.6676767 ,  300.37906054,
       -870.33272996,  492.32141742,  184.85940209,  262.17156228,
        772.18576504,   54.44716253])

In [ ]:
# 절편이 1개
reg.intercept_

150.8602523396147

### 모델 평가

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred = reg.predict(x_test)

In [ ]:
def get_score(y_test, y_pred):
    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_test, y_pred)

    return (mse, rmse, r2)

In [ ]:
mse, rmse, r2 = get_score(y_test, y_pred)
print(mse, rmse, r2)

2712.4894026687316 52.08156490226394 0.39534495486934373


In [ ]:
# 데이터의 크기
n = len(x_train)

# 독립변수의 갯수
k = len(x_train.columns)

# 수정된 결정 계수
adj_r2 = 1 -((1-r2) * (n-1) / (n-k-1))

print('Adjusted R2: {:.4f}'.format(adj_r2))

Adjusted R2: 0.3777


#### 회귀 계수 확인 : 각각의 독립 변수의 설명력 확인

In [ ]:
# 회귀모델에서 구한 회귀계수와 피처 이름을 묶어 시리즈 자료형으로 만들고,
coef = pd.Series(data=reg.coef_, index=x_train.columns)
# 회귀계수값을 기준으로 내림차순으로 정렬
coef.sort_values(ascending=False)

s5     772.185765
bmi    585.667677
s2     492.321417
bp     300.379061
s4     262.171562
s3     184.859402
s6      54.447163
age     42.133330
sex   -256.724253
s1    -870.332730
dtype: float64